### PR TITLE
Initial integration tests for backup controller

### PR DESCRIPTION
**Pull request by jsanda**

### CODE REVIEW

File: controllers/cassandrabackup_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+	cassdcapi "github.com/datastax/cass-operator/operator/pkg/apis/cassandra/v1beta1"
+	api "github.com/k8ssandra/medusa-operator/api/v1alpha1"
+)
+
+const (
+	TestCassandraDatacenterName = "dc1"
+)
+
+var _ = Describe("Reaper controller", func() {
```
Comment (burmanm) [inline controllers/cassandrabackup_controller_test.go:22]: Medusa? ;)

File: controllers/cassandrabackup_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+			{
+				Status: corev1.ConditionTrue,
+				Type:   cassdcapi.DatacenterReady,
+			},
+		}
+		Expect(k8sClient.Status().Patch(context.Background(), cassdc, patch)).Should(Succeed())
+
+		updatedDc := &cassdcapi.CassandraDatacenter{}
+		Expect(k8sClient.Get(context.Background(), cassdcKey, updatedDc)).To(Succeed())
+		Expect(updatedDc.Status.CassandraOperatorProgress).To(Equal(cassdcapi.ProgressReady))
```
Comment (burmanm) [inline controllers/cassandrabackup_controller_test.go:74]: Due to the possible delay of previous creates and updates, should some of these gets do a waiting poll to ensure the previous commands were actually executed by the API server?

File: controllers/suite_test.go [hunk 3]
```diff
@@ -54,26 +56,52 @@ var _ = BeforeSuite(func(done Done) {
...
+	go func() {
+		err = k8sManager.Start(ctrl.SetupSignalHandler())
+		Expect(err).ToNot(HaveOccurred())
+	}()
+
 	k8sClient, err = client.New(cfg, client.Options{Scheme: scheme.Scheme})
 	Expect(err).ToNot(HaveOccurred())
 	Expect(k8sClient).ToNot(BeNil())
 
+	k8sClient = k8sManager.GetClient()
```
Comment (burmanm) [inline controllers/suite_test.go:102]: Confusing, why is the previous k8sClient shadowed here?

File: controllers/cassandrabackup_controller_test.go (hunk 2 above)
Comment (jsanda) [reply controllers/cassandrabackup_controller_test.go:74]: Yeah, I need to add the polling.

File: controllers/suite_test.go (hunk 3 above)
Comment (jsanda) [reply controllers/suite_test.go:102]: Good catch! Just an oversight on my part.

File: backup-client/main.go [hunk 4]
```diff
@@ -101,11 +100,6 @@ func createK8sClient(namespace string) (client.Client, error) {
 }
 
 func doBackup(ctx context.Context, name, addr string) error {
-	if medusaClient, err := medusa.NewClient(addr); err != nil {
-		return err
-	} else {
-		defer medusaClient.Close()
-		return medusaClient.CreateBackup(ctx, name)
-	}
+	//
```
Comment (burmanm) [inline backup-client/main.go:103]: If CreateBackup is no longer called, why does this method exists?

File: controllers/cassandrabackup_controller_test.go [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+				Ports: []corev1.ServicePort{
+					{
+						Name: "cql",
+						Port: 9042,
+					},
+				},
+			},
+		}
+		Expect(k8sClient.Create(context.Background(), dcService)).Should(Succeed())
+		Eventually(func() bool {
```
Comment (burmanm) [inline controllers/cassandrabackup_controller_test.go:98]: These could be ``func() error ....Should(BeNil())``. No need to cast to boolean.

File: controllers/cassandrabackup_controller_test.go [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+				updated.Status.CassdcTemplateSpec.Spec.ServiceAccount == cassdc.Spec.ServiceAccount &&
+				reflect.DeepEqual(updated.Status.CassdcTemplateSpec.Spec.NodeSelector, cassdc.Spec.NodeSelector) &&
+				reflect.DeepEqual(updated.Status.CassdcTemplateSpec.Spec.PodTemplateSpec, cassdc.Spec.PodTemplateSpec) &&
+				reflect.DeepEqual(updated.Status.CassdcTemplateSpec.Spec.Users, cassdc.Spec.Users) &&
+				reflect.DeepEqual(updated.Status.CassdcTemplateSpec.Spec.AdditionalSeeds, cassdc.Spec.AdditionalSeeds)
+		}, timeout, interval).Should(BeTrue())
+
+		By("verify that medusa gRPC clients are invoked")
+		Expect(medusaClientFactory.GetRequestedBackups()).To(Equal(map[string][]string{
+			fmt.Sprintf("192.168.1.50:%d", backupSidecarPort): {backupName},
```
Comment (burmanm) [inline controllers/cassandrabackup_controller_test.go:187]: Are these IP addresses hardcoded somewhere?

File: controllers/cassandrabackup_controller_test.go [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+		}
+		Expect(k8sClient.Create(context.Background(), pod)).Should(Succeed())
+		Eventually(func() bool {
+			created := &corev1.Pod{}
+			err := k8sClient.Get(context.Background(), types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name}, created)
+			return err == nil
+		}, timeout, interval).Should(BeTrue())
+
+		patch := client.MergeFrom(pod.DeepCopy())
+		pod.Status.PodIP = "192.168.1." + strconv.Itoa(50+int(i))
```
Comment (burmanm) [inline controllers/cassandrabackup_controller_test.go:225]: I see, here..

File: backup-client/main.go (hunk 4 above)
Comment (jsanda) [reply backup-client/main.go:103]: `backup-client` is not actually used. I only left it for reference. `CreateBackup` is used [here](http://github.com/k8ssandra/medusa-operator/blob/4bbbcd55d7fbe4d75acd79c9ac18705f75f22494/controllers/cassandrabackup_controller.go#L261-L261)

File: go.mod [hunk 8]
```diff
@@ -3,7 +3,6 @@ module github.com/k8ssandra/medusa-operator
 go 1.13
```
Comment (burmanm) [inline go.mod:3]: Not necessarily this PR, but should we look at updating these builders to use 1.15?

Comment (jsanda) [reply go.mod:3]: Absolutely yes. Let's do it in follow up PRs for both medusa-operator and for reaper-operator. I will create the issues.

File: controllers/cassandrabackup_controller_test.go (hunk 6 above)
Comment (jsanda) [reply controllers/cassandrabackup_controller_test.go]: I added a `getPodIpAddress` function to make it more clear.

File: go.mod (hunk 8 above)
Comment (jsanda) [reply go.mod:3]: @burmanm https://github.com/k8ssandra/medusa-operator/issues/10

**Review by burmanm — APPROVED** [review]

**Pull request merged**